### PR TITLE
"I Belong in CS" teacher promo banner and tile

### PR DESCRIPTION
**Pull request by hannah-whelan**
https://github.com/isaaccomputerscience/isaac-cs-issues/issues/92

### CODE REVIEW

File: src/app/components/elements/cards/NewsCard.tsx
Comment (KasimAliAND) [inline src/app/components/elements/cards/NewsCard.tsx:1]: Nitpicking but we could use logical OR operator instead of ternary operator.
I think this would read a bit better but nothing wrong with your code.
You could do something like this:
`{linkText || "Find out more"}`

File: src/app/components/elements/PromoContent.tsx
Comment (KasimAliAND) [inline src/app/components/elements/PromoContent.tsx:1]: Same as last comment from previous commit. Can use logical OR operator instead of ternary operator on line 24. Personal preference

File: src/app/components/site/Homepage.tsx [hunk 1]
```diff
@@ -56,9 +61,11 @@ export const Homepage = () => {
                                     <ShowMeButtons className={"pt-xl-2"}/>
                                     {/*<img id="homepageHeroImg" className="img-fluid" alt="Three Computer Science students studying with two laptops, one with code on the screen" src="/assets/ics_hero.svg" />*/}
                                 </Col>
-                                <Col data-testid={"featured-news-item"} md="12" lg="7" className="d-none d-lg-block">
+                                {user?.role === "TEACHER" && promo ? (<Col data-testid={"promo-tile"} md="12" lg="7" className="mt-4 d-block">
```
Comment (jmgq) [inline src/app/components/site/Homepage.tsx:64]: This is hard to read and a bit convoluted. A cleaner alternative would be to move this into its own small component (it can live in this file). The component could have the same shape: it would be a `Col` that has another component inside (either TeacherPromoItem or FeaturedNewsItem). So the base shape looks like this:
```jsx
<Col data-testid={"AAAA"} md="12" lg="7" className="BBBB d-lg-block">
    <CCCC item={DDDD} />
</Col>
```
Inside the component you decide what AAAA, BBBB, CCCC and DDDD look like depending on your conditions (which are whether the user is a teacher and there is a promo). But AAAA, BBBB, CCCC and DDDD are all variables.

By doing it like this, you extract this functionality into its own component, which makes this part of the code easier to read, and the component itself encapsulates all the logic that it needs, which makes it easier to understand. In addition, the shape is a lot more clear because rather than repeating it twice (like it's done currently), you just write it once and control how it looks like with variables.

File: src/app/components/site/Homepage.tsx [hunk 2]
```diff
@@ -108,20 +115,32 @@ export const Homepage = () => {
...
+            <section data-testid="promo-content" id="promo-content" className="row bg-primary pattern-05">
+                <ShowLoading
+                until={promoItem}
+                thenRender={()=><Container>
+                    <Col className="py-5">
+                        {promoItem && <PromoContent item={promoItem} />}
+                    </Col>
+                </Container>}
+                />
+                </section></>
```
Comment (jmgq) [inline src/app/components/site/Homepage.tsx:136]: The tabulation here is not right and it makes it harder to read.

File: src/app/components/site/Homepage.tsx [hunk 3]
```diff
@@ -108,20 +115,32 @@ export const Homepage = () => {
...
                 <Container>
                     <Col className="pb-5 pt-4 pattern-04">
                         <h2 className="text-center mb-5">Why sign up?</h2>
                         <WhySignUpTabs/>
                     </Col>
                 </Container>
-            </section>}
+            </section>
+
+            <section data-testid="promo-content" id="promo-content" className="row bg-primary pattern-05">
```
Comment (jmgq) [inline src/app/components/site/Homepage.tsx:127]: You don't need a `data-testid` when you already have an `id`.

File: src/app/components/site/Homepage.tsx [hunk 4]
```diff
@@ -108,20 +115,32 @@ export const Homepage = () => {
...
+                <ShowLoading
```
Comment (jmgq) [inline src/app/components/site/Homepage.tsx:128]: I may not be understanding this correctly, but how come a `ShowLoading` component is needed here? It does not look like `promoItem` ever changes, so I would have thought that you don't need a loading helper.

Comment (jmgq) [reply src/app/components/site/Homepage.tsx:128]: Maybe it's just an alternative way to say: if there is no promo item, just display the loading component. Otherwise, display PromoContent.

File: src/app/components/site/Homepage.tsx (hunk 2 above)
Comment (hannah-whelan) [reply src/app/components/site/Homepage.tsx:136]: This is the usual issue with Prettier not being in use; I can easily format the file to make it look better but that'll make a lot more changes and make the PR harder to spot what actually changed other than indentation.

File: src/test/pages/IsaacApp.test.tsx [hunk 5]
```diff
@@ -139,4 +139,36 @@ describe("IsaacApp", () => {
...
     });
+
+    it('should show the promo content banner for anonymous users', async () => {
+        renderTestEnvironment({role: "ANONYMOUS"});
+        await screen.findByTestId("main");
+        const promoContent = await screen.findByTestId("promo-content");
+        expect(promoContent).toBeInTheDocument();
+    })
+
+    let roles = ["STUDENT", "TUTOR", "TEACHER", "EVENT_LEADER", "EVENT_MANAGER", "CONTENT_EDITOR", "ADMIN"]
```
Comment (jmgq) [inline src/test/pages/IsaacApp.test.tsx:150]: `const roles: UserRole[]`. That way you can get rid of the type cast in `renderTestEnvironment`.

File: src/app/components/elements/cards/NewsCard.tsx
Comment (jmgq) [reply src/app/components/elements/cards/NewsCard.tsx:1]: Another alternative to this is to always display `linkText` and just set "Find out more" as the default parameter value (in the function definition), as this conveys your intention even better.

File: src/app/components/site/Homepage.tsx (hunk 2 above)
Comment (jmgq) [reply src/app/components/site/Homepage.tsx:136]: Yeah... :-( In the meanwhile you can just ensure that the lines you modify have the right indentation. There may be a way to configure your IDE to do this automatically without having to modify the whole file.

File: src/app/components/elements/PromoContent.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            </Col>
+            <Col xs={12} md className="pt-3 pl-3">
+              <h5>
+                <a href={url} target="_blank" rel="noopener noreferrer">
+                  <b>{title}</b>
+                </a>
+              </h5>
+              <div className="text-left">
+                <IsaacContentValueOrChildren
+                  encoding={item.encoding}
```
Comment (jmgq) [inline src/app/components/elements/PromoContent.tsx:37]: You can make `encoding` a variable by adding it to the destructuring in line 9.

File: src/app/components/site/Homepage.tsx (hunk 4 above)
Comment (hannah-whelan) [reply src/app/components/site/Homepage.tsx:128]: This was a funny one, I used the existing ShowLoading component that's used elsewhere (such as FeaturedNewsItem), to let it show Loading until the API request is complete. But it still needed the conditional on line 132 otherwise typescript was unhappy.  

Do you mean it would be better to ignore the ShowLoading component and just do 

`{promoItem ? <PromoContent item={promoItem} /> : <defaultPlaceholder />}`

defaultPlaceholder is the Loading spinner in /handlers/ShowLoading.tsx

I guess though that would eternally show Loading if the API request failed to fetch the content for any reason.

File: src/app/components/elements/PromoContent.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+          <Row className="d-flex align-items-center justify-content-center">
+            <Col
+              xs={12}
+              md={6}
+              lg={5}
+              xl={4}
+              className="d-flex align-items-center justify-content-center"
+            >
+              <img
+                src={path}
```
Comment (jmgq) [inline src/app/components/elements/PromoContent.tsx:25]: `path` can potentially be undefined. What would it happen in that case? Wouldn't we see a broken image in that case?

File: src/app/components/elements/PromoContent.tsx [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import { IsaacPodDTO } from "../../../IsaacApiTypes";
+import { apiHelper } from "../../services";
+import { Col, Row } from "reactstrap";
+import { IsaacContentValueOrChildren } from "../content/IsaacContentValueOrChildren";
+import { Tabs } from "./Tabs";
+
+export const PromoContent = ({ item }: { item: IsaacPodDTO }) => {
+  const { title, subtitle, value, image, url } = item;
+
+  const path = image && image.src && apiHelper.determineImageUrl(image.src);
```
Comment (jmgq) [inline src/app/components/elements/PromoContent.tsx:11]: I think you can simplify this a bit by doing it like this:
```typescript
const path = image?.src && apiHelper.determineImageUrl(image.src);
```

File: src/app/components/elements/PromoContent.tsx [hunk 9]
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+import { IsaacPodDTO } from "../../../IsaacApiTypes";
+import { apiHelper } from "../../services";
+import { Col, Row } from "reactstrap";
+import { IsaacContentValueOrChildren } from "../content/IsaacContentValueOrChildren";
+import { Tabs } from "./Tabs";
+
+export const PromoContent = ({ item }: { item: IsaacPodDTO }) => {
```
Comment (jmgq) [inline src/app/components/elements/PromoContent.tsx:8]: I know that there are some tests in `IsaacApp.test.tsx`, but this new component is missing some unit tests.

File: src/app/components/site/Homepage.tsx (hunk 4 above)
Comment (jmgq) [reply src/app/components/site/Homepage.tsx:128]: Yes, that's exactly what I meant, but I must be misunderstanding something. From a first glance, it looked like this component has no state or side effects (apart from the document's title). That's why I found the use of `ShowLoading` weird: if there is no state or `useEffect` or props, then once the component is loaded, nothing will change, so why do we need this `ShowLoading` component? But I just realised that this file imports 3 things from `state`, so maybe this is using Redux in a way that I'm not familiar with.

TL;DR: ignore me 😂

File: src/test/pages/IsaacApp.test.tsx (hunk 5 above)
Comment (hannah-whelan) [reply src/test/pages/IsaacApp.test.tsx:150]: Changed to 

``` let roles = [
      "STUDENT",
      "TUTOR",
      "TEACHER",
      "EVENT_LEADER",
      "EVENT_MANAGER",
      "CONTENT_EDITOR",
      "ADMIN",
    ] as (UserRole | "ANONYMOUS")[];
```

Due to line 178, "ANONYMOUS" is accepted into renderTestEnvironment, but it isn't a UserRole.

File: src/app/components/site/Homepage.tsx (hunk 1 above)
Comment (hannah-whelan) [reply src/app/components/site/Homepage.tsx]: Refactored this one into `PromoOrFeaturedNews`, now living inside `Dashboard` component to make `Homepage` more readable and less cluttered.

File: src/app/components/elements/PromoContent.tsx (hunk 7 above)
Comment (hannah-whelan) [reply src/app/components/elements/PromoContent.tsx]: path **shouldn't** be undefined, because when creating a newsPod in the content editor it adds an image for you. However just to be extra safe, in case someone decides to delete the image from the content for some reason, I've popped a default image in here.

File: src/app/components/elements/PromoContent.tsx (hunk 9 above)
Comment (hannah-whelan) [reply src/app/components/elements/PromoContent.tsx:8]: Unit tests added for this

File: src/app/components/elements/TeacherPromoItem.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import React from "react";
+import { NewsCard } from "./cards/NewsCard";
+import { IsaacPodDTO } from "../../../IsaacApiTypes";
+import { ShowLoading } from "../handlers/ShowLoading";
+
+export const TeacherPromoItem = ({
```
[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/app/components/elements/TeacherPromoItem.tsx:5]: Unit tests added for this

File: src/mocks/data.ts [hunk 11]
```diff
@@ -4808,6 +4808,34 @@ export const mockNewsPods = {
     totalResults: 5
 };
 
+const promoPodTag = "promo";
```
Comment (jmgq) [inline src/mocks/data.ts:4811]: Unless I'm misunderstanding this, this constant does not add any value: you can simply set the string directly in the only place that is being used. It's clear that this is a promo tag even without reading the variable name.

File: src/mocks/handlers.ts [hunk 12]
```diff
@@ -142,7 +143,15 @@ export const handlers = [
     }),
     rest.get(API_PATH + "/pages/pods/:tag", (req, res, ctx) => {
         const {tag} = req.params;
-        const podsFilteredByTag = produce(mockNewsPods, pods => {
+        let mockPods;
```
Comment (jmgq) [inline src/mocks/handlers.ts:146]: You can simplify this by writing it like this:
```typescript
const mockPods = tag === "promo" ? mockPromoPods : mockNewsPods;
```

If you've used a switch to make it easier to add new tags in the future, then it's better not to do that, because that may never happen and your code is harder to read. It would be a bit of a case of premature optimisation.

File: src/app/components/site/Homepage.tsx [hunk 13]
```diff
@@ -1,173 +1,155 @@
...
-                        <h2 className="h-title mb-4">News</h2>
-                        {user?.loggedIn && <div className={"d-block d-lg-none mb-4 mb-lg-0"}>
-                            <FeaturedNewsItem item={featuredNewsItem} />
-                        </div>}
-                        <NewsCarousel items={carouselNewsItems} />
-                    </div>
-                </Container>
-            </section>
+  if (news) {
+    (user?.loggedIn && user?.role === "TEACHER" && promoItem) ||
```
Comment (jmgq) [inline src/app/components/site/Homepage.tsx:41]: This is a very unconventional way to use the ternary operator. You normally use the ternary operator when you want to assign the result of either branch to some variable. The way you've written it, this may as well be a normal if/else.

File: src/test/pages/IsaacApp.test.tsx [hunk 14]
```diff
@@ -139,4 +139,59 @@ describe("IsaacApp", () => {
...
+      "EVENT_LEADER",
+      "EVENT_MANAGER",
+      "CONTENT_EDITOR",
+      "ADMIN",
+    ] as (UserRole | "ANONYMOUS")[];
+
+    it.each(roles)(
+      "should not show the promo content banner for %s users",
+      async (role) => {
+        renderTestEnvironment({ role: role as UserRole });
```
Comment (jmgq) [inline src/test/pages/IsaacApp.test.tsx:163]: You shouldn't need the casting anymore.

File: src/mocks/handlers.ts (hunk 12 above)
Comment (hannah-whelan) [reply src/mocks/handlers.ts:146]: Yeah I was trying to keep it so that it would be easily useable for further tags, but I'll simplify as suggested!

Comment (jmgq) [reply src/mocks/handlers.ts]: I understand. In these cases, unless you can guarantee that changes are definitely going to happen imminently (like in the current or next sprint), it's better to just keep it simple. Because otherwise what ends up happening in the vast majority of cases is that the use case you were foreseeing never actually happens, and the result is that the code is harder to read than it needed to be.

File: src/test/components/elements/Dashboard.test.tsx [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+const mockPromoItem = mockPromoPods.results[0];
+const mockFeaturedNewsItem = mockNewsPods.results[1];
+
+describe("Dashboard", () => {
+  const setupTest = (role: UserRole | "ANONYMOUS", props = {}) => {
+    renderTestEnvironment({
+      role: role,
+      PageComponent: Dashboard,
+      componentProps: {
+        featuredNewsItem: mockFeaturedNewsItem as IsaacPodDTO,
```
Comment (jmgq) [inline src/test/components/elements/Dashboard.test.tsx:16]: I don't think these castings are needed (the same goes for the other one in `PromoContent.test.tsx`). I'll approve the PR anyway, since it's quite minor.

**Review by jmgq — APPROVED** [review]

File: src/test/components/elements/Dashboard.test.tsx (hunk 15 above)
Comment (hannah-whelan) [reply src/test/components/elements/Dashboard.test.tsx]: I thought it was required since the component was expecting props as a certain type, but it seems to work without  🎉

Comment (jmgq) [reply src/test/components/elements/Dashboard.test.tsx]: I think that as long as the object has the expected shape, it does not need to implement an interface / type. Duck typing.
